sinks/splunk/splunk.go

```diff
@@ -0,0 +1,154 @@
+package splunk
+
+import (
+	"context"
+	"crypto/tls"
+	"net/http"
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	hec "github.com/fuyufjh/splunk-hec-go"
+	"github.com/sirupsen/logrus"
+	"github.com/stripe/veneur/protocol"
+	"github.com/stripe/veneur/sinks"
+	"github.com/stripe/veneur/ssf"
+	"github.com/stripe/veneur/trace"
+	"github.com/stripe/veneur/trace/metrics"
+)
+
+type splunkSpanSink struct {
+	*hec.Client
+	hostname    string
+	sendTimeout time.Duration
+
+	// Total counts of sent and dropped spans, respectively
+	sentCount, dropCount uint32
+
+	traceClient *trace.Client
+	log         *logrus.Logger
+}
+
+// NewSplunkSpanSink constructs a new splunk span sink from the server
+// name and token provided, using the local hostname configured for
+// veneur. An optional argument, validateServerName is used (if
+// non-empty) to instruct go to validate a different hostname than the
+// one on the server URL.
+func NewSplunkSpanSink(server string, token string, localHostname string, validateServerName string, log *logrus.Logger, sendTimeout time.Duration) (sinks.SpanSink, error) {
+	client := hec.NewClient(server, token).(*hec.Client)
+
+	if validateServerName != "" {
+		tlsCfg := &tls.Config{}
+		tlsCfg.ServerName = validateServerName
+
+		trnsp := &http.Transport{TLSClientConfig: tlsCfg}
+		httpC := &http.Client{Transport: trnsp}
+		client.SetHTTPClient(httpC)
+	}
+
+	return &splunkSpanSink{Client: client, hostname: localHostname, log: log, sendTimeout: sendTimeout}, nil
+}
+
+func (sss *splunkSpanSink) Start(cl *trace.Client) error {
+	sss.traceClient = cl
+	return nil
+}
+
+func (sss *splunkSpanSink) Flush() {
+	samples := &ssf.Samples{}
+	samples.Add(
+		ssf.Count(
+			sinks.MetricKeyTotalSpansFlushed,
+			float32(atomic.SwapUint32(&sss.sentCount, 0)),
+			map[string]string{"sink": sss.Name()}),
+		ssf.Count(
+			sinks.MetricKeyTotalSpansDropped,
+			float32(atomic.SwapUint32(&sss.dropCount, 0)),
+			map[string]string{"sink": sss.Name()},
+		),
+	)
+
+	metrics.Report(sss.traceClient, samples)
+	return
+}
+
+// Name returns this sink's name
+func (*splunkSpanSink) Name() string {
+	return "splunk"
+}
+
+// Ingest takes in a span and passes it to Splunk using the
+// HTTP Event Collector
+func (sss *splunkSpanSink) Ingest(ssfSpan *ssf.SSFSpan) error {
```

Comment:
Yeah, definitely intentional. Since the single-event api is much nicer to use (it lets us set metadata for each event sent to the HEC), and not flushing lets us not consume memory we probably need to consume (HEC is surprisingly reasonably-sped!)

Reply:
Yes, we definitely want to support batching at scale. One HTTP connection per envoy span is a recipe for problems in a large environment! Though we started simple.